### PR TITLE
 Pre-check some video IDs in the archive before downloading

### PR DESCRIPTION
**Pull request by jbruchon**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
    This tries to stop single video downloads that are in the archive
    file from triggering any actual HTTP requests. This became an
    issue during the playlist download capability being broken by the
    disable_polymer option; as a workaround, I put together download
    batch files using a regex-manipulated copy of the playlist page
    DOM, so each batch file would run a download for each single video
    in a playlist rather than having the program fetch the playlist.

    The major problem is that even if a video ID is in the archive,
    the video info page must still be downloaded before it is ever
    checked within the archive for rejection. By checking the video ID
    before anything is downloaded, the rejection happens much faster
    and no unnecessary HTTP requests are sent out.

    This should be extended in the future for other services that do
    not require a page download to retrieve the video ID.

    This uses a regex match to check the site-specific URL and should
    basically be a no-op for all other sites.

### CODE REVIEW

**Comment by jbruchon** [comment]
I also just fixed a problem caused by UPX being in the PATH resulting in a corrupt executable from `make_win.bat`.

**Comment by pukkandan** [comment]
It's a great idea. But I think the actually checking should be done by the extractor instead of in the main module

**Comment by jbruchon** [comment]
@pukkandan That'd be nice, but due to the way the program is constructed, the archive info and the YoutubeDL.py functions don't seem to be available to the extractors.

**Comment by pukkandan** [comment]
What I meant is that it would be best if the extractor can be made to provide the id instead of hardcoding the urls in YouTubeDL

**Comment by jbruchon** [comment]
@pukkandan I agree...here's where the youtube extractor pulls the ID: https://github.com/blackjack4494/yt-dlc/blob/master/youtube_dlc/extractor/youtube.py#L1802-L1806

The problem is this: how do you then get that ID back up to the main code? I'm primarily used to working with C, and this program is basically a pile of object-oriented spaghetti, so I don't know how I'm supposed to kick that back up. It takes forever just to follow this program's flow.

**Comment by pukkandan** [comment]
This works, at least by the minimal testing I have done https://github.com/blackjack4494/yt-dlc/pull/181

Since it affects all extractors, It would need proper testing before it can actually be used

**Comment by pukkandan** [comment]
btw, you should make the make_win fix a seperate PR

**Comment by jbruchon** [comment]
@pukkandan I reverted make_win and set up a separate PR (I accidentally committed it to the wrong branch in the first place). The code in this PR should affect no other extractors since it requires a regex domain match and I'm unaware of any other extractors that will use `https://*.[youtube|bitchute].com/`. I don't like that I had to add the code where I did, but I did a lot of looking for a better place to put it and always found that "the cure is worse than the disease." I'm sure there's a better way to write this code, I just can't think of it (and my itch is now scratched, too.)

**Comment by jbruchon** [comment]
@pukkandan Take another look now; the extractor is known by the time the code is called, so I made the code explicitly check the extractor rather than relying on a regex match. This should be safer and faster too, and effectively makes this feature a no-op for any extractor that isn't implemented within it.

**Comment by pukkandan** [comment]
@jbruchon Did you not see the PR I mentioned above? #181 

I have already made generalized code for all (or most) extractors, without needing regex defined in YoutubeDL

> #181 works, at least by the minimal testing I have done
Since it affects all extractors, It would need proper testing before it can actually be used
